### PR TITLE
[Workspace] feat: Add workspace navigation for default route

### PR DESCRIPTION
**Pull request by yubonluo**
### Description
Add workspace navigation for default route to handle the following cases:
 - If user has a default workspace configured, go to overview page of that workspace
 - If user only has one workspace, go to overview page of that workspace
 - If user has more than one workspaces, go to homepage
 - If user has no workspaces, go to initial page (initial is a new page)

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7783
## Screenshot

https://github.com/user-attachments/assets/aaeadefe-1ea7-42ee-87af-e2e47c00f292

## Changelog

- feat: [Workspace] Add workspace navigation for default route

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/public/plugin.ts [hunk 1]
```diff
@@ -374,6 +375,18 @@ export class WorkspacePlugin
       workspaceAvailability: WorkspaceAvailability.outsideWorkspace,
     });
 
+    // register workspace navigation
+    core.application.register({
+      id: WORKSPACE_NAVIGATION_APP_ID,
+      title: '',
+      navLinkStatus: AppNavLinkStatus.hidden,
+      async mount(params: AppMountParameters) {
+        const { renderNavigationApp } = await import('./application');
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/plugin.ts:384]: Actually we can just navigate to the desired app without rendering a page. That should speed up the redirect.

File: src/plugins/workspace/server/plugin.ts [hunk 2]
```diff
@@ -115,12 +116,16 @@ export class WorkspacePlugin implements Plugin<WorkspacePluginSetup, WorkspacePl
...
           { request, logger: this.logger },
-          { page: 1, perPage: 1 }
+          { page: 1, perPage: 2 }
         );
+        const basePath = core.http.basePath.serverBasePath;
+
         if (workspaceListResponse?.success && workspaceListResponse.result.total > 0) {
-          return toolkit.next();
+          return response.redirected({
+            headers: { location: `${basePath}/app/${WORKSPACE_NAVIGATION_APP_ID}` },
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/plugin.ts:125]: We should have know which workspace to go in server side, we could just get the right url to go into the specific workspace.

File: src/plugins/workspace/public/components/workspace_navigation_app.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        (workspace) => workspace.id === defaultWorkspaceId
+      );
+      // If user has a default workspace configured, go to overview page of that workspace
+      // If user has more than one workspaces, go to homepage
+      if (defaultWorkspace) {
+        navigateToWorkspaceOverview(availableUseCases, defaultWorkspace, application, http);
+      } else {
+        application.navigateToApp('home');
+      }
+    }
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_navigation_app.tsx:61]: I'd prefer to do these in server side, the application in browser side is only used to detect which feature is the most top features of the specific workspace.

File: src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx [hunk 4]
```diff
@@ -69,7 +73,7 @@ export const WorkspaceInitial = () => {
             defaultMessage: 'Create a workspace',
           })}
           description={
-            <EuiToolTip content={isDashboardAdmin || noAdminToolTip}>
+            <EuiToolTip content={isDashboardAdmin ? <></> : noAdminToolTip}>
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx:76]:
```suggestion
            <EuiToolTip content={isDashboardAdmin ? null : noAdminToolTip}>
```
What about we just render null?

File: src/plugins/workspace/public/plugin.ts (hunk 1 above)
Comment (yubonluo) [reply src/plugins/workspace/public/plugin.ts]: Sure, the code has been updated

File: src/plugins/workspace/server/plugin.ts (hunk 2 above)
Comment (yubonluo) [reply src/plugins/workspace/server/plugin.ts]: Sure, the code has been updated

File: src/plugins/workspace/public/components/workspace_navigation_app.tsx (hunk 3 above)
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_navigation_app.tsx]: Sure, I have deleted the render code.

File: src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx (hunk 4 above)
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx]: updated

File: src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx [hunk 5]
```diff
@@ -40,10 +40,14 @@ export const WorkspaceInitial = () => {
...
-      'Contact your administrator to create a workspace or to be added to an existing one.',
-  });
+  const noAdminToolTip = (
+    <>
+      {i18n.translate('workspace.initial.card.createWorkspace.toolTip', {
+        defaultMessage:
+          'Contact your administrator to create a workspace or to be added to an existing one.',
+      })}
+    </>
+  );
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx:50]: Why was this change necessary?

File: src/plugins/workspace/public/plugin.test.ts [hunk 6]
```diff
@@ -40,7 +40,7 @@ describe('Workspace plugin', () => {
       savedObjectsManagement: savedObjectManagementSetupMock,
       management: managementPluginMock.createSetupContract(),
     });
-    expect(setupMock.application.register).toBeCalledTimes(5);
+    expect(setupMock.application.register).toBeCalledTimes(6);
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/plugin.test.ts:43]: Do you, by chance, have an explanation for this number? It would help to have a comment... unless it is obvious.

File: src/plugins/workspace/server/plugin.test.ts [hunk 7]
```diff
@@ -205,31 +224,97 @@ describe('Workspace server plugin', () => {
...
+    });
+
+    it('with / request path and default workspace', async () => {
+      const request = httpServerMock.createOpenSearchDashboardsRequest({
+        path: '/',
+      });
+      const savedObjectsClient: jest.Mocked<SavedObjectsClientContract> = savedObjectsClientMock.create();
+      const mockUiSettings = {
+        get: jest.fn().mockResolvedValue('default'),
+      };
```
Comment (AMoo-Miki) [inline src/plugins/workspace/server/plugin.test.ts:287]: For future-proofing, this should use `uiSettingsServiceMock` and not reinvent it.

File: src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx (hunk 5 above)
Comment (yubonluo) [reply src/plugins/workspace/public/components/workspace_initial/workspace_initial.tsx:50]: It has been solved in other pr(https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7771)

File: src/plugins/workspace/server/plugin.ts [hunk 8]
```diff
@@ -115,12 +116,44 @@ export class WorkspacePlugin implements Plugin<WorkspacePluginSetup, WorkspacePl
       if (path === '/') {
         const workspaceListResponse = await this.client?.list(
           { request, logger: this.logger },
-          { page: 1, perPage: 1 }
+          { page: 1, perPage: 100 }
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/plugin.ts:119]:
```suggestion
          { page: 1, perPage: 2 }
```
I guess 2 should be good enough?

File: src/plugins/workspace/server/plugin.ts [hunk 9]
```diff
@@ -115,12 +116,44 @@ export class WorkspacePlugin implements Plugin<WorkspacePluginSetup, WorkspacePl
...
+        const basePath = core.http.basePath.serverBasePath;
+
         if (workspaceListResponse?.success && workspaceListResponse.result.total > 0) {
-          return toolkit.next();
+          const workspaceList = workspaceListResponse.result.workspaces;
+          // If user only has one workspace, go to overview page of that workspace
+          if (workspaceList.length === 1) {
+            return response.redirected({
+              headers: {
+                location: `${basePath}/w/${workspaceList[0].id}/app/${WORKSPACE_NAVIGATION_APP_ID}`,
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/server/plugin.ts:129]: Maybe we can try using the `formatUrlWithWorkspaceId` method.

File: src/plugins/workspace/server/plugin.ts [hunk 10]
```diff
@@ -115,12 +116,44 @@ export class WorkspacePlugin implements Plugin<WorkspacePluginSetup, WorkspacePl
...
+          // If user has more than one workspaces, go to homepage
+          if (defaultWorkspace) {
+            return response.redirected({
+              headers: {
+                location: `${basePath}/w/${defaultWorkspace.id}/app/${WORKSPACE_NAVIGATION_APP_ID}`,
+              },
+            });
+          } else {
+            return response.redirected({
+              headers: { location: `${basePath}/app/home` },
```
Comment (Hailong-am) [inline src/plugins/workspace/server/plugin.ts:152]: i see this logic depends on `core.http.registerOnPostAuth`, how about if security is disabled case, will it work?

File: src/plugins/workspace/server/plugin.ts (hunk 9 above)
Comment (yubonluo) [reply src/plugins/workspace/server/plugin.ts:129]: In the server side, the `formatUrlWithWorkspaceId ` method will be useless, because the parameter `basePath: IBasePath` is public side.

File: src/plugins/workspace/server/plugin.ts (hunk 10 above)
Comment (yubonluo) [reply src/plugins/workspace/server/plugin.ts:152]: Although the security plugin is disable, the `core.http.registerOnPostAuth` will also be triggered.

File: src/plugins/workspace/public/plugin.test.ts (hunk 6 above)
Comment (yubonluo) [reply src/plugins/workspace/public/plugin.test.ts]: Sure, I have added a annotate.

File: src/plugins/workspace/server/plugin.test.ts (hunk 7 above)
Comment (yubonluo) [reply src/plugins/workspace/server/plugin.test.ts]: Sure, I have updated the code

File: src/plugins/workspace/public/plugin.ts [hunk 11]
```diff
@@ -375,6 +377,32 @@ export class WorkspacePlugin
       workspaceAvailability: WorkspaceAvailability.outsideWorkspace,
     });
 
+    const registeredUseCases$ = this.registeredUseCases$;
+    // register workspace navigation
```
Comment (ruanyl) [inline src/plugins/workspace/public/plugin.ts:381]: Would be nice to have a more context about this change

**Review by ruanyl — APPROVED** [review]

File: src/plugins/workspace/server/plugin.ts (hunk 9 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/plugin.ts:138]: I c, thanks for clarification.

**Review by SuZhou-Joe — APPROVED** [review]
